rpc/server_interceptors.go

```diff
@@ -0,0 +1,126 @@
+package rpc
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/edaniels/golog"
+	"go.opencensus.io/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// UnaryServerTracingInterceptor starts a new Span if Span metadata exists in the context.
+func UnaryServerTracingInterceptor(logger golog.Logger) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		remoteSpanContext, err := remoteSpanContextFromContext(ctx)
+		var span *trace.Span
+		if err == nil {
+			ctx, span = trace.StartSpanWithRemoteParent(ctx, "server_root", remoteSpanContext)
+			defer span.End()
+		} else {
+			logger.Warnf("client did not send valid Span metadata in headers, local Spans will not be linked to client. reason: %w", err)
+		}
+
+		resp, err := handler(ctx, req)
+		if err == nil {
+			return resp, nil
+		}
+		if _, ok := status.FromError(err); ok {
+			return nil, err
```

Comment:
Rather than leaving the status info out, which might be helpful, you could wrap the error with it. Something like
```
import "github.com/pkg/errors"
...
		if s, ok := status.FromError(err); ok {
			return nil, errors.Wrapf(err, s.Message())
		}


```